sympy/stats/joint_rv.py

```diff
@@ -162,3 +207,115 @@ def marginal_distribution(rv, *indices):
     if hasattr(prob_space.distribution, 'marginal_distribution'):
         return prob_space.distribution.marginal_distribution(indices, rv.symbol)
     return prob_space.marginal_distribution(*indices)
+
+
+class CompoundDistribution(Basic, NamedArgsMixin):
+    """
+    Represents a compound probability distribution.
+
+    Constructed using a single probability distribution with a parameter
+    distributed according to some given distribution.
+    """
+    def __new__(cls, dist):
+        assert isinstance(dist, (ContinuousDistribution, DiscreteDistribution))
```

Comment:
Might be beneficial for end-user to get a a better explicit error message if instantiation is attempted to be done from an illegal distribution.
Such as:
```
if not isinstance(dist, (ContinuousDistribution, DiscreteDistribution)):
    raise ValueError('CompoundDistribution can only be initialized from ContinuousDistribution or DiscreteDistribution')
```
or something similar.